common/app/views/support/cleaner/AmpEmbedCleaner.scala

```diff
@@ -60,6 +60,33 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
 
   }
 
+  object AmpSoundcloud {
+    def createElement(document: Document, trackId: String): Element = {
+      val soundcloud = document.createElement("amp-soundcloud")
+      soundcloud.attr("data-trackid", trackId)
+      soundcloud.attr("data-visual", "true")
+      soundcloud.attr("height", "300") // height is necessary if data-visual == true
+    }
+
+    def getTrackIdFromUrl(soundcloudUrl: String): String = {
+      val pattern = "soundcloud.com%2Ftracks%2F(\\d+)".r
+      pattern.findAllIn(soundcloudUrl).matchData.map { matches => matches.group(1) }.mkString
+    }
+
+    def clean(document: Document) = {
+      document.getElementsByClass("element-audio").foreach { audioElementWithIframe: Element =>
+        audioElementWithIframe.getElementsByTag("iframe").foreach { iframeElement: Element =>
+          val soundcloudUrl = iframeElement.attr("src")
+          Option(getTrackIdFromUrl(soundcloudUrl)) filterNot { _.isEmpty } foreach { trackId =>
+            val soundcloudElement = createElement(document, trackId)
+            audioElementWithIframe.appendChild(soundcloudElement)
+            iframeElement.remove()
+          }
```

Comment:
the structure is good, but to simplify this glue logic you can use a for comprehension I think - as all the things are seq/option.  Also you can split the pure stuff and the mutating a little more if you like.  Something along these lines perhaps:
```
// work out what we need to change
val toDo = for {
  audioElementWithIframe<- document.getElementsByClass("element-audio")
  iframeElement <- audioElementWithIframe.getElementsByTag("iframe")
  soundcloudUrl = iframeElement.attr("src")
  trackId <- getTrackIdFromUrl(soundcloudUrl)
} yield (audioElementWithIframe, iframeElement, createElement(document, trackId))
// go through and do the mutation to the jsoup dom
toDo.foreach { case (audioElementWithIframe, iframeElement, soundcloudElement) =>
  audioElementWithIframe.appendChild(soundcloudElement)
  iframeElement.remove()
}
```